[IRGen] Fix ABI for thick async functions.

nate-chandler opened the pull request:
Previously, thick async functions were represented sometimes as a pair of (AsyncFunctionPointer, nullptr)--when the thick function was produced via a thin_to_thick_function, e.g.--and sometimes as a pair of (FunctionPointer, ThickContext)--when the thick function was produced by a partial_apply--with the size stored in the slot of the ThickContext.

That optimized for the wrong case: partial applies of dynamic async functions; in that case, there is no appropriate AsyncFunctionPointer to form when lowering the partial_apply instruction.  The far more common case is to know exactly which function is being partially applied.  In that case, we can form the appropriate AsyncFunctionPointer.

Furthermore, the previous representation made calling a thick function more complex: it was always necessary to check whether the context was in fact null and then proceed along two different paths depending.

Here, that behavior is corrected by creating a thunk in a mandatory IRGen SIL pass in the case that the function that is being partially applied is dynamic.  That new thunk is then partially applied in place of the original partial_apply of the dynamic function.

slavapestov commented:
Can we ban partial applies of non-constant functions in SIL? They should only be emitted in a few narrow cases and it would simplify IRGen to not have to deal with this.

nate-chandler commented:
To be safe, we probably don't want to change the behavior for sync functions at the moment.

Does SILGen ever emit partial applies of non-constant async functions?  If not, it would make sense to add a check in the verifier that no partial apply is of a function that is non-constant and async.

Until we have more async code to test on though, it may be tough to be entirely confident that no partial applies of non-constant async functions will ever appear.  Creating a thunk if one does will serve as a backup against that possibility.

Moving partial apply forwarder creation out of IRGen would address the root issue of IRGen complexity around partial applies, including the case of partial applies of non-constant functions.

nate-chandler committed:
[Test] Gave async/partial_apply[_forwarder].sil pa'd funcs sizes.

Previously, the partially applied functions were not defined and as such
had no sizes.  That resulted in downstream issues as an attempt was made
to replace the size in the AsyncFunctionPointer for the functions'
forwarders with the size in the AsyncFunctionPointer for functions
because the latter did not exist.

nate-chandler committed:
[Test] Added for part_ap of async @conv(method).

nate-chandler committed:
[Test] FC'd part_ap of async dyn witness_method.

nate-chandler committed:
[Test] Tested @callee_guaranteed witness_method.

Previously, the test verified the behavior the partial application of
of dynamic async @convention(witness_method) functions which produced
@callee_owned thick functions.  That callee convention is no longer
used, so here that test is changed to verify the behavior of producing
@callee_guaranteed thick functions which are in fact still used.

nate-chandler committed:
[Test] Updated FC'd PAF names.

Now that thunks are introduced for async dynamic functions which are
partially applied, the application forwarders of those thunks will have
different names.

nate-chandler committed:
[AST] De/mangling for functions with generic sigs.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[Test] Passed -g in IRGen/async/partial_apply.sil.

nate-chandler committed:
[Gardening] Fixed sigil.

nate-chandler committed:
[IRGen] Added LinkEntity for partial apply forwarder.

Up to now, there had been no need to define a LinkEntity for a partial
apply forwarder.  Now that async partial apply forwarders will each have
their own async function pointer, an entity is needed to pass to the
code that generates the async function pointers.

No demangling or remangling changes are required because that code has
existed for as long as partial apply forwarders to support demangling
their symbols.

nate-chandler committed:
[Gardening] Tweaked comment.

nate-chandler committed:
[IRGen] Added LinkEntity for AFP to PAF.

The new AFP entity follows the pattern set by previous AFP entities.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please clean test linux platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please smoke test macos platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test linux platform

aschwaighofer commented:
@swift-ci please test os x

aschwaighofer commented:
@swift-ci please test linux platform

aschwaighofer commented:
The failures are due to the new commit:

```
[AST] NFC: Make ExtInfo param Optional<>  
```

aschwaighofer commented:
You probably need to update the `FunctionType::get` calls to include an `ExtInfo`

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler commented:
@swift-ci please clean test macos platform

nate-chandler commented:
@swift-ci please test windows platform

aschwaighofer commented:
The lldb failure should get fixed when https://github.com/apple/llvm-project/pull/2666 merges

aschwaighofer commented:
@swift-ci Please clean test os x

nate-chandler committed:
[Test] Mangling of second PAF for a function.

nate-chandler committed:
[IRGen] Fix ABI for thick async functions.

Previously, thick async functions were represented sometimes as a pair
of (AsyncFunctionPointer, nullptr)--when the thick function was produced
via a thin_to_thick_function, e.g.--and sometimes as a pair of
(FunctionPointer, ThickContext)--when the thick function was produced by
a partial_apply--with the size stored in the slot of the ThickContext.

That optimized for the wrong case: partial applies of dynamic async
functions; in that case, there is no appropriate AsyncFunctionPointer to
form when lowering the partial_apply instruction.  The far more common
case is to know exactly which function is being partially applied.  In
that case, we can form the appropriate AsyncFunctionPointer.

Furthermore, the previous representation made calling a thick function
more complex: it was always necessary to check whether the context was
in fact null and then proceed along two different paths depending.

Here, that behavior is corrected by creating a thunk in a mandatory
IRGen SIL pass in the case that the function that is being partially
applied is dynamic.  That new thunk is then partially applied in place
of the original partial_apply of the dynamic function.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler commented:
@swift-ci please test windows platform

aschwaighofer reviewed (approved):
looks good to me

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci Please test Apple Silicon macOS Platform
